### DISCUSSION TITLE
PageProps not getting updated on client side navigation

### DISCUSSION THREAD

**Discussion by HiteshDatt** [vercel/next.js#40521] [Help] [answered]
I noticed this behavior with my Nextjs app (ISR). let's say a user lands on the homepage, and then navigates to the about-us page(using Nextjs Link), on the first visit I can see a fetch request in the Network tab for the about-us page (about-us.json), but after that however many times the user change routes, that fetch call is not made again, this way unless the user does a full reload of the website, the user will not see the updated about-us page.
**Updating the user's app without having the user reload the full website** is a critical requirement for my SPA, how can I go about handling this situation?

**Comment by gaston-flores** [comment]
Did you set `revalidate` in your getStaticProps? Without it, the page props will be generated only once at build time. This example works for me:

```jsx
import Link from "next/link";

export default function AboutPage({ time }) {
  return (
    <div>
      About us - {time}{" "}
      <Link href="/">
        <a>Go back</a>
      </Link>
    </div>
  );
}

export async function getStaticProps() {
  return {
    props: {
      time: Date.now(),
    },
    revalidate: 10,
  };
}
```

And I see the requests in my network tab

![image](https://user-images.githubusercontent.com/18354222/190044535-a6b3a56a-8eae-4d55-8d8b-d42aec8ef522.png)

Can you share an example with the issue?

**Reply by HiteshDatt** [reply]
Yes, I'm using revalidate, and the page props are updating fine only on the reload of the page. But if a user has landed on the page already and then makes navigation using Nextjs Link( which does not reloads the full page) you will not see the request for the page in the network tab after the first navigation. In your example above it looks like you're reloading the page, can you please try without reloading the page and only using Nextjs Link to navigate from homepage back to about page a few times, you'll not see the request again in the network tab unless you do a full reload of the page.

**Reply by HiteshDatt** [reply]
@gaston-flores here I have a 50-second-long video showing the exact behavior and the problem associated with it, Please go through it whenever you can spare a minute or two, thanks [https://www.youtube.com/watch?v=qCqiB8Kce9w](https://www.youtube.com/watch?v=qCqiB8Kce9w)

**Reply by gaston-flores** [reply]
That's very strange. The screenshot I shared is using `next/link` for client side navigation. Can you try the same snippet in a fresh `create-next-app` install?

**Reply by HiteshDatt** [reply]
Yep, I did try it out on a fresh app, tried it on total 3 nextjs apps, basically it is expected behaviour in nextjs apps, **it caches the prefetched data for pages, I need a way to somehow bust this cache without having the user reload** the full website, in which case the cache is already busted

**Comment by PsyGik** [comment]
That's the expected behavior for ISR. Page re-validation will only occur when a new page request is made to the server. ~Client-side page navigation are cached and do not trigger re-validation~. You can use [On-demand re-validation](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#on-demand-revalidation) or do [client side data fetching](https://nextjs.org/docs/basic-features/data-fetching/client-side#client-side-data-fetching-with-swr).

See: https://github.com/vercel/next.js/discussions/40521#discussioncomment-3651945

**Reply by HiteshDatt** [reply]
Hi thanks for going through the issue, as you said Page re-validation will only occur when a new page request is made to the server is definitely the right behavior, but I was talking about the scenario when there are multiple users on the website, and the page is already fresh in the nextjs server, still there's no way for the user using the website to get the updated pageProps in their browser, the only way for the browser(client) to get the data which is present in the nextjs server is by a full reload, which is a big problem for single page applications (especially those trying to make a PWA out of it). 
Other than this using client side data fetching does not work for us as getting the prerendered pages was the only benefit for which we moved to nextjs

**Reply by PsyGik** [reply]
You can use `<a>` without `<Link>` for your case. Or look into one of the [router APIs](https://nextjs.org/docs/api-reference/next/router#routerreload)

**Reply by PsyGik** [reply]
So this peeked my interest and I tried to replicate your issue. Here's a sample repo that shows how ISR works. https://github.com/PsyGik/try-nextjs-link-isr-issue

Given a `revalidate` prop, `next/link` does indeed trigger a revalidate, and on the next navigation the page is updated.

**Reply by HiteshDatt** [reply]
Hey @PsyGik Thanks a lot! Found the issue, the issue is with the nextjs version :( the functionality is completely changing only on the basis of the nextjs version, look at this [https://stackblitz.com/edit/nextjs-jrq6ev?file=package.json](https://stackblitz.com/edit/nextjs-jrq6ev?file=package.json) I only changed the next version to "12.1.5".
So, to conclude, in version "12.1.5" it behaves like how I described above, and in version "12.3.0" it is behaving how you described and how I want it to behave, thanks a lot!

**Reply by sdjnes** [reply]
Hello! Does this still work for you? In both the examples above after running `build` and `start`, going between the about page and homepage with next/link doesn't trigger the revalidation. I only see a single console log and only by clicking the `a` link will it revalidate.

**Reply by HiteshDatt** [reply]
> Hello! Does this still work for you? In both the examples above after running `build` and `start`, going between the about page and homepage with next/link doesn't trigger the revalidation. I only see a single console log and only by clicking the `a` link will it revalidate.

Yes, make sure your "next" version is "12.3.0" in the package.json, the Link component fetches the page data everytime that link is hovered on in this version

**Comment by HiteshDatt** [comment, marked as answer]
To conclude, in version "12.1.5" of next the Link tag behaves in the way that the page data is (pre)fetched only when the link for the page comes in the viewport, and in version "12.3.0" the Link tag fetches the page data every time the link is hovered on.
